src/Membership.sol

```diff
@@ -0,0 +1,360 @@
+    /// @param index the index of the membership in the membership set
+    event MembershipExpired(uint256 idCommitment, uint32 membershipRateLimit, uint32 index);
+
+    /// @notice Emitted when a membership is erased by its holder during grace period
+    /// @param idCommitment the idCommitment of the membership
+    /// @param membershipRateLimit the rate limit of this membership
+    /// @param index the index of the membership in the membership set
+    event MembershipErased(uint256 idCommitment, uint32 membershipRateLimit, uint32 index);
+
+    /// @notice Emitted when a membership in its grace period is extended (i.e., is back to Active state)
+    /// @param idCommitment the idCommitment of the membership
+    /// @param membershipRateLimit the rate limit of this membership
+    /// @param index the index of the membership in the membership set
+    /// @param newGracePeriodStartTimestamp the new grace period start timestamp of this membership
+    event MembershipExtended(
+        uint256 idCommitment, uint32 membershipRateLimit, uint32 index, uint256 newGracePeriodStartTimestamp
+    );
+
+    /// @dev contract initializer
+    /// @param _priceCalculator Address of an instance of IPriceCalculator
+    /// @param _maxTotalRateLimit Maximum total rate limit of all memberships in the membership set
+    /// @param _minMembershipRateLimit Minimum rate limit of each membership
+    /// @param _maxMembershipRateLimit Maximum rate limit of each membership
+    /// @param _activeDurationForNewMemberships Active state duration of each membership
+    /// @param _gracePeriodDurationForNewMemberships Grace period duration of each membership
+    function __MembershipUpgradeable_init(
+        address _priceCalculator,
+        uint32 _maxTotalRateLimit,
+        uint32 _minMembershipRateLimit,
+        uint32 _maxMembershipRateLimit,
+        uint32 _activeDurationForNewMemberships,
+        uint32 _gracePeriodDurationForNewMemberships
+    )
+        internal
+        onlyInitializing
+    {
+        __MembershipUpgradeable_init_unchained(
+            _priceCalculator,
+            _maxTotalRateLimit,
+            _minMembershipRateLimit,
+            _maxMembershipRateLimit,
+            _activeDurationForNewMemberships,
+            _gracePeriodDurationForNewMemberships
+        );
+    }
+
+    function __MembershipUpgradeable_init_unchained(
+        address _priceCalculator,
+        uint32 _maxTotalRateLimit,
+        uint32 _minMembershipRateLimit,
+        uint32 _maxMembershipRateLimit,
+        uint32 _activeDurationForNewMemberships,
+        uint32 _gracePeriodDurationForNewMemberships
+    )
+        internal
+        onlyInitializing
+    {
+        require(0 < _minMembershipRateLimit);
+        require(_minMembershipRateLimit <= _maxMembershipRateLimit);
+        require(_maxMembershipRateLimit <= _maxTotalRateLimit);
+        require(_activeDurationForNewMemberships > 0);
```

Comment:
Perhaps we should follow the linter's advice and add error messages here?

Reply:
Not sure if worth it. These are checks that will be executed only during deployment. It's highly likely that we'll be using values hardcoded in the deploy script that will never hit these `requires` (like we do now).